src/server/pfs/server/source.go

```diff
@@ -51,10 +53,16 @@ func NewSource(commitInfo *pfs.CommitInfo, fs fileset.FileSet, opts ...SourceOpt
 		s.dirIndexOpts = append(s.dirIndexOpts, index.WithRange(&index.PathRange{
 			Lower: sc.pathRange.Lower,
 		}))
+		// The upper for the index path range is set to be past the provided upper
+		// to ensure that directories between the last file that should be emitted
+		// and the first file that shouldn't get created.
+		// For example, the files /d1/f1 and /d2/f2 with a path range of [/d1/f1, /d2/f2) should
+		// emit /d1/f1 and /d2/.
 		s.fileIndexOpts = append(s.fileIndexOpts, index.WithRange(&index.PathRange{
 			Lower: sc.pathRange.Lower,
-			Upper: sc.pathRange.Upper,
+			Upper: sc.pathRange.Upper + "_",
```

Comment:
If we don't explode on invalid paths, I would prefer to add a null byte here.  I prefer the guarantee that there is no path between `sc.pathRange.Upper` and the index.PathRange upper bound. I believe the guarantee here is that there is no currently valid path.  If we change the allowed character set, this could break.

Reply:
It wouldn't break with an allowable character set change because the upper bound is still used in the source iteration, but using the null byte is probably clearer anyways.